### DISCUSSION TITLE
How to install specified version of LLVM/Clang?

### DISCUSSION THREAD

**Discussion by lost22git** [termux/termux-packages#22051] [Q&A]
I want to install llvm-18, but installed package only llvm-19.

**Comment by ghost** [comment]
If you need downgrading, you probably need to compile the package yourself. You could also clone this repo, create a new branch and push a commit that reverts libllvm to 18 and a github action should start. Once finished, there will be an available `.deb` file which can be installed with, `dpkg`. Note that lib incompatibilies may arise from downgrading libllvm and I will not be liable.

**Reply by Biswa96** [reply]
You can install older llvm from TUR https://github.com/termux-user-repository/tur/tree/master/tur but not llvm-18 yet.

**Reply by lost22git** [reply]
> If you need downgrading, you probably need to compile the package yourself. You could also clone this repo, create a new branch and push a commit that reverts libllvm to 18 and a github action should start. Once finished, there will be an available `.deb` file which can be installed with, `dpkg`. Note that lib incompatibilies may arise from downgrading libllvm and I will not be liable.

I did it and got an error:

![Screenshot_20241104-051738__01](https://github.com/user-attachments/assets/9034bf87-636e-433c-a96c-045cd572fedc)
I revert to this commit:
```
❯ git log
commit b4261d9858061f00a71c58449e2af276a811dae6 (HEAD -> llvm-18, origin/llvm-18)
Author: Chongyun Lee <45286352+licy183@users.noreply.github.com>
Date:   Thu Jun 27 09:47:13 2024 +0800

    libllvm: fix build (#20685)
```

**Reply by ghost** [reply]
You're suppose to commit on top of the `main` upstream, not resetting to where llvm-18 was commited.

```
# retrieve upstream updates
git checkout main && git pull
git checkout -b llvm-18
# apply patch for llvm 18
git apply <(git diff -R b4261d9858061f00a71c58449e2af276a811dae6 packages/libllvm)
# apply patch for flang 18
git apply <(git diff -R 24875893e65985e54639736a61cc8a9c8abd6ea0 packages/flang)
# commit and push
```

See https://github.com/harieamjari/termux-packages/actions/runs/11665193459
